src/pymor/algorithms/bfgs.py

```diff
@@ -0,0 +1,320 @@
+
+    if gradient_mu_coefficient > 0.:
+        hessian_grad_diff = np.dot(hessian, gradient_difference)
+        quadratic_coefficient = np.dot(gradient_difference, hessian_grad_diff)
+
+        hessian_part = np.outer(hessian_grad_diff, mu_difference) + np.outer(mu_difference, hessian_grad_diff)
+        mu_part = np.outer(mu_difference, mu_difference)
+
+        mu_coeff = (gradient_mu_coefficient + quadratic_coefficient) / gradient_mu_coefficient**2
+        hessian_coeff = 1. / gradient_mu_coefficient
+
+        hessian += mu_coeff * mu_part - hessian_coeff * hessian_part
+    else:
+        hessian = np.eye(hessian.shape[0])
+
+    return hessian
+
+
+@defaults('miniter', 'maxiter', 'rtol', 'tol_sub', 'stagnation_window', 'stagnation_threshold')
+def error_aware_bfgs(model, parameter_space=None, initial_guess=None, miniter=0, maxiter=100, rtol=1e-16,
+         tol_sub=1e-8, line_search_params=None, stagnation_window=3, stagnation_threshold=np.inf,
+         error_aware=False, error_criterion=None, beta=None, radius=None, return_stages=False):
+    """BFGS algorithm.
+
+    This method solves the optimization problem ::
+
+        min J(mu), mu in C
+
+    for an output functional depending on a box-constrained `mu` using the BFGS method.
+
+    In contrast to `scipy.optimize` with the `L-BFGS-B` methods, this BFGS implementation is
+    explicitly designed to work with an error estimator. In particular, this implementation
+    terminates if the higher level TR boundary from `pymor.algorithms.tr` is reached instead of
+    continuing to optimize close to the boundary.
+
+    Parameters
+    ----------
+    model
+        The |Model| used for the optimization.
+    parameter_space
+        If not `None`, the |ParameterSpace| for enforcing the box constraints on the
+        parameter `mu`. Otherwise a |ParameterSpace| with no constraints.
+    initial_guess
+        If not `None`, a |Mu| instance of length 1 containing an initial guess for the
+        solution `mu`. Otherwise, a random parameter from the parameter space is chosen
+        as the initial value.
+    miniter
+        Minimum amount of iterations to perform.
+    maxiter
+        Fail if the iteration count reaches this value without converging.
+    rtol
+        Finish when the relative error measure is below this threshold.
+    tol_sub
+        Finish when the clipped parameter error measure is below this threshold.
+    line_search_params
+        Dictionary of additional parameters passed to the line search method.
+    stagnation_window
+        Finish when the parameter update has been stagnating within a tolerance of
+        `stagnation_threshold` during the last `stagnation_window` iterations.
+    stagnation_threshold
+        See `stagnation_window`.
+    error_aware
+        If `True`, perform an additional error aware check during the line search phase.
+        Intended for use with the trust region algorithm. Requires the parameters `beta`
```

Comment:
Is there documentation somewhere, what to pass for the additional parameters in this case?